keps/sig-autoscaling/2702-graduate-hpa-api-to-GA/README.md

```diff
@@ -0,0 +1,426 @@
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+###### How can someone using this feature know that it is working for their instance?
+
+<!--
+For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
+for each individual pod.
+Pick one more of these and delete the rest.
+Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
+and operation of this feature.
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+- [ ] Events
+  - Event Reason: 
+- [ ] API .status
+  - Condition name: 
+  - Other field: 
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+<!--
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
+  - Metric name:
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+<!--
+Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+implementation difficulties, etc.).
+-->
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### Does this feature depend on any specific services running in the cluster?
+
+<!--
+Think about both cluster-level services (e.g. metrics-server) as well
+as node-level agents (e.g. specific version of CRI). Focus on external or
+optional services that are needed. For example, if this feature depends on
+a cloud provider API, or upon an external software-defined storage or network
+control plane.
+
+For each of these, fill in the following—thinking about running existing user workloads
+and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]
+    - Usage description:
+      - Impact of its outage on the feature:
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+### Scalability
```

Comment:
> Why is the Scalability section N/A?

@johnbelamaric I was thinking along the lines of this graduation will not change the scalability characteristics of the v2beta2 API.  But let me try and address the scalability changes from v1 to v2:

```
The HPA v2 APIs allow users to configure multiple metrics, each with a separate target.  A recommendation is
calculated for each metric and the largest recommendation is used.  The more metrics are added to a given
HPA the longer it will take to reconcile.  The HPA is single-threaded processing recommendations one-at-a-time.
When default reconciliation period is 15 seconds.  If there is too much work to do reconciliation will slow down 
and happen less frequently than every 15 seconds.  This will cause autoscaling to be less responsive at high scale.

Previously v1 scaled along two dimensions, number of HPA and number of pods selected by each HPA (linearly).  
Now it will scale with the number of metrics defined in HPAs and the number of pods selected each metric (again 
linearly).

Additionally v2 adds a behavior structure which allows the user configure that rate and delay of scaling and down.
  Enforcing these constraints require storing previous recommendations and scaling events in memory.  The 
longer the configured interval the more memory is used.  The maximum window allows is 60 minutes 
([code](https://pkg.go.dev/k8s.io/api/autoscaling/v2beta2#HPAScalingRules)) so 240 recommendations / events per 
configured metric.  Each recommendation is an `int32` and `time.Time`.  Each scaling event is an `int32`, 
a `time.Time` and a `bool` ([code](https://pkg.go.dev/k8s.io/api/autoscaling/v2beta2#HPAScalingRules)) so 
the memory footprint is relatively small.  It will scale linearly with the number of metrics defined and the size 
of the HPA's configured window.
```